### PR TITLE
Add tests for RFmx driver resource management

### PR DESCRIPTION
**Pull request by astarche**
### What does this Pull Request accomplish?

Add tests for RFmx driver resource management.

Fixes [AB#1781546](https://ni.visualstudio.com/94b22d7b-ad7b-4f5e-88f0-867910f91c94/_workitems/edit/1781546).

### Why should this Pull Request be merged?

Ensure that RFmx driver session management works OK with grpc-device session management and provide examples of how they interact.

### What testing has been done?

Ran and passed all RFmxSpecAn tests.

### CODE REVIEW

**Comment by astarche** [comment]
image

File: source/tests/system/nirfmxspecan_driver_api_tests.cpp [hunk 1]
```diff
@@ -424,6 +445,200 @@ TEST_F(NiRFmxSpecAnDriverApiTests, BuildSpurString_ReturnsSpurString)
...
+    }
+
+    return param_value;
+  }
+};
+
+using ResourcePair = std::tuple<std::string, std::string>;
+
+template <typename T>
+inline std::vector<T> vector_concat(const std::vector<T>& first, const std::vector<T> second)
```
Comment (reckenro) [inline source/tests/system/nirfmxspecan_driver_api_tests.cpp:491]: Minor: should we pass `second` in by reference as well?

File: source/tests/system/nirfmxspecan_driver_api_tests.cpp [hunk 2]
```diff
@@ -424,6 +445,200 @@ TEST_F(NiRFmxSpecAnDriverApiTests, BuildSpurString_ReturnsSpurString)
...
+}
+
+// Same as NiRFmxSpecAnDriverApiSuccessfulResourceInitTests, but with 3 sessions instead of the standard 2.
+TEST_F(NiRFmxSpecAnDriverApiTests, InitializeThreeIdenticalResourcesAsSeparateSessions_CloseEachSession_EachIsClosedIndependently)
+{
+  const auto session1 = init_session(stub(), PXI_5663, "specan1,specan2");
+  const auto session2 = init_session(stub(), PXI_5663, "specan2,specan1");
+  const auto session3 = init_session(stub(), PXI_5663, "specan1,specan2");
+  EXPECT_NE(session1.id(), session2.id());
+  EXPECT_NE(session1.id(), session3.id());
```
Comment (reckenro) [inline source/tests/system/nirfmxspecan_driver_api_tests.cpp:556]: Should we go ahead and add `EXPECT_NE(session2.id(), session3.id())`? Just to make sure we're not getting {1, 2, 2} or something for the IDs.

**Review by reckenro — APPROVED** [review]

**Review by mgumble — APPROVED** [review]
Awesome nice to have a good example of parameterized tests too

File: source/tests/system/nirfmxspecan_driver_api_tests.cpp [hunk 3]
```diff
@@ -424,6 +445,201 @@ TEST_F(NiRFmxSpecAnDriverApiTests, BuildSpurString_ReturnsSpurString)
...
+};
+
+// Implements a function object that can be used as the name generator
+// for parameterized test suites instantiated from NiRFmxSpecAnDriverApiResourceInitTests.
+class ResourceTuplePrinter {
+ public:
+  using ParamType = NiRFmxSpecAnDriverApiResourceInitTests::ParamType;
+  std::string operator()(const TestParamInfo<ParamType>& info)
+  {
+    return sanitize_param<0>(info.param) + "__" + sanitize_param<1>(info.param);
```
Comment (mgumble) [inline source/tests/system/nirfmxspecan_driver_api_tests.cpp:469]: May be worth adding a little special logic for the empty string params.
image

**Pull request merged**